providers/yaml/src/main/java/org/jboss/resteasy/plugins/providers/YamlProvider.java

```diff
@@ -110,4 +141,110 @@ public void writeTo(Object t, Class<?> type, Type genericType, Annotation[] anno
+         // Get the TCCL first
+         ClassLoader result = Thread.currentThread().getContextClassLoader();
+         if (result == null) {
+            result = type.getClassLoader();
+         }
+         return result;
+      });
+   }
+
+   private static Collection<String> createAllowedList() {
+      final String value;
+      if (System.getSecurityManager() == null) {
+         value = System.getProperty(ALLOWED_LIST);
+      } else {
+         value = AccessController.doPrivileged((PrivilegedAction<String>) () -> System.getProperty(ALLOWED_LIST));
+      }
+      return value == null ? Collections.emptyList() : Arrays.asList(value.split(","));
+   }
+
+   private static boolean isTypeCheckDisabled() {
+      final String value;
+      if (System.getSecurityManager() == null) {
+         value = System.getProperty(DISABLE_TYPE_CHECK);
+      } else {
+         value = AccessController.doPrivileged((PrivilegedAction<String>) () -> System.getProperty(DISABLE_TYPE_CHECK));
+      }
+      return value != null && (value.isEmpty() || value.equalsIgnoreCase("true"));
+   }
+
+   private static class TypeSafeConstructor extends CustomClassLoaderConstructor {
+      private final Set<Class<?>> types;
+      private final Collection<String> allowedList;
+
+      private TypeSafeConstructor(final ParameterizedType parameterizedType, final ClassLoader classLoader,
+                                  final Collection<String> allowedList) {
+         super(classLoader);
+         this.allowedList = allowedList;
+         final Set<Class<?>> genericTypes = new HashSet<>();
+         for (Type typeArg : parameterizedType.getActualTypeArguments()) {
+            try {
+               genericTypes.add(classLoader.loadClass(typeArg.getTypeName()));
+            } catch (ClassNotFoundException e) {
+               LogMessages.LOGGER.failedToLoadType(typeArg.getTypeName());
+            }
+         }
+         this.types = genericTypes;
+      }
+
+      private TypeSafeConstructor(final Class<?> type, final ClassLoader classLoader,
+                                  final Collection<String> allowedList) {
+         super(classLoader);
+         this.types = Collections.singleton(type);
+         this.allowedList = allowedList;
+      }
+
+      @Override
+      protected Object newInstance(final Class<?> ancestor, final Node node, final boolean tryDefault) {
+         if (denied(node.getType())) {
+            throw Messages.MESSAGES.typeNotAllowed(node.getType());
+         }
+         return super.newInstance(ancestor, node, tryDefault);
+      }
+
+      private boolean denied(final Class<?> type) {
+         if (type == null) {
+            return false;
+         }
+         final String name = type.getName();
+         // Check the allowed types first
+         if (allowedList.contains(name)) {
+            return false;
+         }
+         // Allow all primitives
+         if (type.isPrimitive()) {
+            return true;

```

Comment:
If you want to allow all primitives, shouldn't it be "false" here?

Reply:
Yes, nice catch.